Library/OcDeviceMiscLib/ForgeUefi.c

```diff
@@ -93,14 +110,34 @@ OcForgeUefiSupport (
     return EFI_INVALID_PARAMETER;
   }
 
-  NewBS = AllocateZeroPool (sizeof (EFI_BOOT_SERVICES));
-  if (NewBS == NULL) {
-    DEBUG ((DEBUG_INFO, "OCDM: Failed to allocate BS copy\n"));
-    return EFI_OUT_OF_RESOURCES;
+  if (Trash) {
+    Signature = *(UINT64 *)(&gBS->CreateEventEx);
```

Comment:
Does old Apple EFI allocate gBS and gDS consecutively and this checks for it? Can't you use the header size or something?

Reply:
The layout is old Apple EFI, but it is directly from standard edk EFI code, in fact.

What is being checked for here is that the memory about to be trashed is - fortuitously - exactly that value: something which is only used when the memory is being loaded, in fact (when we check for references to this value throughout edk code).

The point being, I _have_ to trash exactly that QWORD of memory. I would have just trashed it and prayed... but in the targeted firmware I can confirm that it is harmless, and proceed.

Does this sound okay, given that explanation?

Reply:
Yes, thanks. Is this commented somewhere and I missed it?

Reply:
No it's not - I'll add a comment. (I get the idea the existing OC code style is "don't comment if it's obvious what it's doing", where 'obvious' tends to mean, "read it 10 times and scratch your head for a week" - so I try not to over-comment! ;) )

Reply:
No, the existing code style is “You have two weeks off-ish? Write a Mach-O dynamic linker, fix most of the bugs, integrate it with a separate prelinkedkernel mod logic into a working kext injection and fuzz it” :) When the deadline is not yesterday, I strongly prefer comments, I think I tried with the crypto lib